FileSystemSimulator/FileSystem/FileSystem.cpp

```diff
@@ -0,0 +1,317 @@
+
+  OpenFileTable::OFTEntry* entry = oft->getEntry(i_index);
+  if (!entry)
+    return false;
+
+  FileDescriptor fd = getFileDescriptor(entry->fd_index);
+  if (fd.file_size == 0) {
+    bool success = allocateDataBlock(entry->fd_index);
+    if (!success)
+      return false;
+  }
+
+  while (true) {
+
+    size_t buffer_pos = entry->cur_pos % Sector::BLOCK_SIZE;
+    size_t buffer_space = Sector::BLOCK_SIZE - buffer_pos;
+
+    if (i_count < buffer_space) {
+      std::memcpy(entry->buffer + buffer_pos, i_mem_area, i_count);
+      entry->cur_pos += i_count;
+      return true;
+    }
+
+    std::memcpy(entry->buffer + buffer_pos, i_mem_area, buffer_space);
+    size_t block_number = fd.data_blocks[entry->cur_pos / Sector::BLOCK_SIZE];
+    iosystem->write_block(block_number, entry->buffer);
+    entry->cur_pos += buffer_space;
+    i_count -= buffer_space;
+
+    if (entry->cur_pos >= fd.file_size) {
+      fd.file_size = entry->cur_pos;
+      writeFileDescriptorToIO(fd, entry->fd_index);
+
+      bool success = allocateDataBlock(entry->fd_index);
+      if (!success)
+        return false;
+    }
+  }
+}
+
+int FileSystem::findFreeFileDescriptor()
+{
+  size_t first_fd_index = BITMAP_BLOCKS_NUMBER;
+
+  for (size_t i = first_fd_index; i < FIRST_DATA_BLOCK_INDEX; i++) {
+    char block[Sector::BLOCK_SIZE];
+    iosystem->read_block(i, block);
+    for (size_t j = 0; j < FDS_IN_BLOCK; j++) {
+      FileDescriptor* fd = (FileDescriptor*)block + j;
+      if (fd->is_free)
+        return (i - first_fd_index)*FDS_IN_BLOCK + j;
+    }
+  }
+
+  return -1;
+}
+
+bool FileSystem::recordFileToDir(const std::string & i_file_name, size_t i_fd_index)
+{
+  FileDescriptor dir_fd = getFileDescriptor(0);
+  size_t entries_number = dir_fd.file_size / sizeof(DirEntry);
+  size_t entries_in_last_block = entries_number % ENTRIES_IN_BLOCK;
+
+  // search for existing file with name i_file_name
+  int first_free_index = -1;
+  for (int i = 0; i <= dir_fd.getLastBlockIndex(); i++) {
+    char block[Sector::BLOCK_SIZE];
+    iosystem->read_block(dir_fd.data_blocks[i], block);
+
+    for (size_t j = 0; j < ENTRIES_IN_BLOCK; j++) {
+      DirEntry* entry = (DirEntry*)block + j;
+      if (strcmp(entry->file_name, i_file_name.c_str()) == 0)
+        return false;
+
+      // check if this entry is free
+      if (strcmp(entry->file_name, "") == 0 && first_free_index == -1)
+        first_free_index = i * ENTRIES_IN_BLOCK + j;
+
+      if (i*ENTRIES_IN_BLOCK + j == entries_number - 1)
+        break;
+    }
+  }
+  if (first_free_index == -1) {
+    first_free_index = entries_number;
+
+    dir_fd.file_size += sizeof(DirEntry);
+    writeFileDescriptorToIO(dir_fd, 0);
+  }
+
+  // add one more data block for dir file if needed
+  if (entries_in_last_block == 0 && first_free_index != -1) {
+    bool success = allocateDataBlock(0);
+    if (!success)
+      return false;
+    // update dir_fd after allocation
+    dir_fd = getFileDescriptor(0);
+  }
+
```

Comment:
Review the following checks, be sure that the file size is not increased in vain

Reply:
created issues